### PR TITLE
Improve type coercions for paramWhiteList

### PR DESCRIPTION
**Pull request by LMS007**
- Rework coercions so they work more like react props rather than function calls
- Improve type-coercion capability to allow for nullish values. This is similar to propTypes `.isRequired`
- Add coercions to remaining config settings host-config.

partially fixes: #1968

-----------

Example usage..

```
  const paramWhiteList = {
    annotations: coercions.toString.orNull,
    group: coercions.toString.orNull,
    query: coercions.toString.orNull,
    appType: coercions.toString.orNull,
    openSidebar: coercions.toBoolean,
    ...
```
When reviewing this, the diff is hard to follow and makes little sense because the `type-coercions.js` is totally restructured. 

Image diff (master is on the right, this branch on the left) --it's a screenshot of the host-config after its run through the  `paramWhiteList`. They are identical.

Screen Shot 2020-09-12 at 2 49 55 PM

### CODE REVIEW

File: src/shared/type-coercions.js [hunk 1]
```diff
@@ -55,15 +114,61 @@ export function toObject(value) {
...
+  return toObjectShapeShim;
+};
+
+// Exported coercion types
+const coercions = {
+  toString: toStringShim,
+  toInteger: toIntegerShim,
+  toObject: toObjectShim,
+  toBoolean: toBooleanShim,
+  custom: customShim,
```
Comment (LMS007) [inline src/shared/type-coercions.js:170]: toObject.shape and custom are similar, the former will ensure the object is in fact an object before the custom shape is called, whereas custom take the raw value of any time. I felt both were different enough to warrant both. Currently, toObject.shape is not used anywhere, but I suspect we will use it if we extend this.

File: src/shared/type-coercions.js [hunk 2]
```diff
@@ -55,15 +114,61 @@ export function toObject(value) {
...
+  any: v => v,
```
Comment (LMS007) [inline src/shared/type-coercions.js:171]: This is more or less throwing my hands up because the values need to be re-thunk'ed a bit. 
Perfect example is `showHighlights` where it is either a specific string or a boolean. I think it means we need to refactor the API and then force this into a more strict type of one or the other. Worst case, `showHighlights` could be a custom if we need to still support legacy values.

File: src/sidebar/test/host-config-test.js [hunk 3]
```diff
@@ -65,9 +65,9 @@ describe('hostPageConfig', function () {
     assert.deepEqual(hostPageConfig(window_), {});
   });
 
-  it('ignores `null` values in config', function () {
+  it('ignores `undefined` values in config', function () {
```
Comment (LMS007) [inline src/sidebar/test/host-config-test.js:68]: a `null` value does now have meaning so only `undefind` will be ignored.

**Comment by robertknight** [comment]
Rebased on master.

**Review by robertknight — COMMENTED** [review]
Hi @LMS007 - I finally got around to looking at this, sorry for the delay.

The question at the top of my mind here is whether we want to silently _coerce_ values of the "wrong" type or _validate_ them and simply ignore values of the wrong type.

IIRC the original motivation for introducing coercion of client settings was to make it easier to support passing additional configuration from Via to the client, with fewer changes to Via itself.

However a consequence of this is that if a publisher accidentally passes an incorrect value, we will silently accept it and it may or may not do what they expect. We are then required to leave with that loose handling of settings forever - since removing it could potentially break existing web pages.

I have a general preference for being conservative in what is accepted because it is always easier to relax constraints than to introduce them later. Let's catch up about it in Slack.
